patches/tModLoader/Terraria/Social/Base/SocialBrowserModule.cs

```diff
@@ -0,0 +1,206 @@
+	internal ModDownloadItem[] CachedInstalledModDownloadItems { get; set; }
+
+	public ModDownloadItem[] DirectQueryInstalledMDItems(QueryParameters qParams = new QueryParameters()) {
+		var mods = GetInstalledMods();
+		var listIds = new List<ModPubId_t>();
+
+		foreach (var mod in mods) {
+			if (GetModIdFromLocalFiles(mod.modFile, out var id))
+				listIds.Add(id);
+		}
+
+		qParams.searchModIds = listIds.ToArray();
+
+		return DirectQueryItems(qParams);
+	}
+
+	public ModDownloadItem[] GetInstalledModDownloadItems()
+	{
+		if (CachedInstalledModDownloadItems == null) {
+			CachedInstalledModDownloadItems = DirectQueryInstalledMDItems();
+		}
+
+		return CachedInstalledModDownloadItems;
+	}
+
+	/////// Specialty Internal LocalMod related Methods ///////////////////////////////////////////
+	public bool DoesAppNeedRestartToReinstallItem(ModPubId_t modId);
+
+	internal bool DoesItemNeedUpdate(ModPubId_t modId, LocalMod installed, Version webVersion);
+
+	internal IReadOnlyList<LocalMod> GetInstalledMods();
+
+	internal LocalMod IsItemInstalled(string slug)
+	{
+		return GetInstalledMods().Where(t => t.Name == slug).FirstOrDefault();
+	}
+
+	/////// Management of Downloads ///////////////////////////////////////////
+
+	/// <summary>
+	/// Downloads all UIModDownloadItems provided.
+	/// </summary>
+	internal Task SetupDownload(List<ModDownloadItem> items, int previousMenuId)
+	{
+		// Can't update enabled items due to in-use file access constraints
+		// Mod unloading needs to occur off the main thread o avoid UI impacts/soft-freeze
+		var needFreeInUseMods = items.Any(item => { item.UpdateInstallState(); return item.IsEnabled; });
+		if (needFreeInUseMods) {
+			Interface.loadMods.suppressAutoLoad = true;
+			Main.menuMode = Interface.loadModsID;
+			var unload = Task.Run(() => ModLoader.ModLoader.Unload());
+			return unload.ContinueWith(delegate { DownloadRunner(items, previousMenuId, needFreeInUseMods); });
+		}
+		else {
+			return DownloadRunner(items, previousMenuId, needFreeInUseMods);
+		}
+	}
+
+	private Task DownloadRunner(List<ModDownloadItem> items, int previousMenuId, bool needReload)
+	{
+		IDownloadProgress progress = null;
+
+		if (!Main.dedServ) {
+			// Create UIWorkshopDownload
+			var ui = new UIWorkshopDownload(previousMenuId);
+			Main.MenuUI.SetState(ui);
+			progress = ui;
+		}
+
+		return Task.Run(() => InnerDownload(progress, items, needReload));
+	}
+
+	private void InnerDownload(IDownloadProgress uiProgress, List<ModDownloadItem> items, bool reloadWhenDone)
+	{
+		var changedModsSlugs = new HashSet<string>();
+
+		foreach (var item in items) {
+			DownloadItem(item, uiProgress);
+
+			// Due to issues with Steam moving files from downloading folder to installed folder,
+			// there can be some latency in detecting it's installed. Fine tune if it's giving issues - Solxan
+			Thread.Sleep(2500);
```

Comment:
don't suppose we can do this at the end of downloading multiple mods? Or do some sort of check?